lms/services/canvas_studio.py

```diff
@@ -86,11 +86,11 @@ class File(TypedDict):
     id: str
     display_name: str
 
-    duration: float | None
+    duration: NotRequired[float | None]
```

Comment:
Alternatively we could include `None` in the dicts created below.

Reply:
The [corresponding frontend types](https://github.com/hypothesis/lms/blob/c7d7d9afdd9d39ca071555c261ac44903af1ad5a/lms/static/scripts/frontend_apps/api-types.ts#L30) are:

```ts
export type File = FileBase & {
  type: 'File';

  /**
   * MIME type of the file.
   *
   * This can either be a top-level type (eg. "video", "text") or a sub-type
   * ("text/html").
   */
  mime_type?: MimeType;

  /** URL of a thumbnail for this item. */
  thumbnail_url?: string;

  /** Duration of audio or video in seconds. */
  duration?: number;
}
```

In TS the `?` after the name is shorthand for adding `| undefined` to the type, which I believe is equivalent to `NotRequired` here. So to match the frontend, it would be `duration: NotRequired[float]`.

JS is unfortunately more complex than TS because there are more ways to indicate "not present" in an object:

1. Omit the field
2. Provide the field, with a value of `undefined`
3. Provide the field, with a value of `null`

Whereas in Python you'd have only:

1. Property is missing
2. Property is present with a value of `None`

For a given property, it is preferable to pick one way to represent absent, either `foo: T | null` or `foo?: T`. While various conditional JS language constructs treat `null` and `undefined` interchangeably, not all of them do.

Reply:
Went for NotRequired[T] which I think matched the frontend expectations and typechecks in the BE.